ports/esp32: Added Ethernet Support.

eydam-prototyping opened the pull request:
Ethernet-PHYs from ESP-IDF (LAN8720, IP101, RTL8201, DP83848) are now
supported.

Signed-off-by: Tobias Eydam <eydam-prototyping@outlook.com>

eydam-prototyping committed:
ports/esp32: Added Ethernet Support.

Ethernet-PHYs from ESP-IDF (LAN8720, IP101, RTL8201, DP83848) are now
supported.

Signed-off-by: Tobias Eydam <eydam-prototyping@outlook.com>

eydam-prototyping commented:
Ethernet Support for ESP32
```python
import network
import machine
import time

print("Initializing LAN")
l = network.LAN(mdc = machine.Pin(23), mdio = machine.Pin(18), phy_type = network.PHY_LAN8720, phy_addr=0)

while l.status() != network.ETH_INITIALIZED:
    print(".")
    time.sleep(1)

print("Done. Starting LAN")
l.active(True)

while (l.status() != network.ETH_STARTED) & (l.status() != network.ETH_CONNECTED):
    print(".")
    time.sleep(1)

print("Done. Waiting for IP")

while l.status() != network.ETH_GOT_IP:
    print(".")
    time.sleep(1)

print("Got IP:      {}".format(l.ifconfig()[0]))
print("    Subnet:  {}".format(l.ifconfig()[1]))
print("    Gateway: {}".format(l.ifconfig()[2]))
print("    DNS:     {}".format(l.ifconfig()[3]))
print("    MAC:     {}".format(":".join(["{:02x}".format(b) for b in l.config("mac")])))
```

Output: 
```bash
Initializing LAN
Done. Starting LAN
.
Done. Waiting for IP
.
.
.
.
Got IP:      192.168.178.70
    Subnet:  255.255.255.0
    Gateway: 192.168.178.1
    DNS:     192.168.178.1
    MAC:     24:0a:c4:9e:ec:3f
```
I tested it with LAN8710, but I don't have the other chips to test. 

Some checks are failing, but I have no Idea why. 
I can compile it with ESP-IDF v4.0.2, v4.2, v4.3.

This is my first PR, so if I can do something better, just let me know.

dpgeorge commented on ports/esp32/modnetwork.c:
```diff
@@ -753,17 +753,17 @@ STATIC mp_obj_t esp_phy_mode(size_t n_args, const mp_obj_t *args) {
     { MP_ROM_QSTR(MP_QSTR_WLAN), MP_ROM_PTR(&get_wlan_obj) },
-    #if !MICROPY_ESP_IDF_4
+    
+    #if (ESP_IDF_VERSION_MAJOR == 4) && (ESP_IDF_VERSION_MINOR >= 2)
```
So does LAN only work with IDF v4.2 or greater?  What about v4.02 and v4.1.1?

eydam-prototyping replied:
I checked it: it works with v4.1 and greater. 
esp_netif.h and some functions are new in version 4.1.
I changed it and left a comment.

dpgeorge commented on ports/esp32/modnetwork.c:
```diff
@@ -784,18 +784,21 @@ STATIC const mp_rom_map_elem_t mp_module_network_globals_table[] = {
     { MP_ROM_QSTR(MP_QSTR_PHY_IP101), MP_ROM_INT(PHY_IP101) },
+    { MP_ROM_QSTR(MP_QSTR_PHY_RTL8201), MP_ROM_INT(PHY_RTL8201) },
+    { MP_ROM_QSTR(MP_QSTR_PHY_DP83848), MP_ROM_INT(PHY_DP83848) },
+    #if ESP_IDF_VERSION_MINOR >= 3
```
We no longer support IDF v3, so this line can be removed.

eydam-prototyping replied:
This is the minor version number. 
PHY_KSZ8041 is new in version 4.3
I left a comment in the code

dpgeorge commented on ports/esp32/modnetwork.c:
```diff
@@ -784,18 +784,21 @@ STATIC const mp_rom_map_elem_t mp_module_network_globals_table[] = {
+    { MP_ROM_QSTR(MP_QSTR_ETH_STOPPED), MP_ROM_INT(ETH_STOPPED)},
+    { MP_ROM_QSTR(MP_QSTR_ETH_CONNECTED), MP_ROM_INT(ETH_CONNECTED)},
+    { MP_ROM_QSTR(MP_QSTR_ETH_DISCONNECTED), MP_ROM_INT(ETH_DISCONNECTED)},
+    { MP_ROM_QSTR(MP_QSTR_ETH_GOT_IP), MP_ROM_INT(ETH_GOT_IP)},
```
It would be nice to reuse the constants below for the WLAN class, because that makes it more consistent (LAN and WLAN are conceptually similar) and also more portable to other ports (non esp32 ports).

eydam-prototyping replied:
The other constants are named STAT_... which is specific for WLAN, as far as I know. And I don't want to change the existing constants. 
Do you think I should change it anyway?

dpgeorge replied:
Let's leave it as it is.

dpgeorge commented:
Thank you, this is a great contribution!

dpgeorge commented:
Please also check the CI failures: code formatting and esp32 build with IDF v4.3 (and esp32s2).

eydam-prototyping committed:
ports/esp32: Added Ethernet Support.

Added some comments and correced the code formatting.
Fixed supported targets: ESP32S2 (not supported)

Signed-off-by: Tobias Eydam <eydam-prototyping@outlook.com>

eydam-prototyping commented:
CI checks are passing now.
Only the question concerning the constants is still open (see above).

chtinnes commented on ports/esp32/network_lan.c:
```diff
@@ -116,63 +119,88 @@ STATIC mp_obj_t get_lan(size_t n_args, const mp_obj_t *pos_args, mp_map_t *kw_ar
+            self->phy = esp_eth_phy_new_dp83848(&phy_config);
+            break;
+        case PHY_KSZ8041:
+            self->phy = esp_eth_phy_new_ksz8041(&phy_config);
```
This will create a warning for IDF 4.1

eydam-prototyping replied:
Ok, thank you. KSZ8041 is new in Version 4.3. I will change that.

chtinnes commented:
I've built this with ESP-IDF release/v4.2.

```
import network
import machine
l = network.LAN(mdc = machine.Pin(23), mdio = machine.Pin(18), phy_type = network.PHY_LAN8720, phy_addr=0)
```

It results in 

`OSError: esp_eth_driver_install() failed.`

Any ideas what can go wrong?

eydam-prototyping commented:
What hardware do you use? Are you sure mdc and mdio pins are set correctly? I tested it with Olimex ESP32-EVB

chtinnes commented:
@eydam-prototyping: I have my own Hardware which is basically the same as the WaveShare board using the wiring here: https://sautter.com/blog/ethernet-on-esp32-using-lan8720/

With the IDF-3 version it worked fined but I used `phy_addr=1`.

eydam-prototyping committed:
ports/esp32/network_lan.c: More detailed error output.

esp_eth_dirver_install gives more details on failure.
PHY_KSZ8041 only for ESP-IDF 4.3 and higher.
argument phy_addr will now be used.

Signed-off-by: Tobias Eydam <eydam-prototyping@outlook.com>

eydam-prototyping commented:
You are right, I forgot to use phy_addr. The argument was checked, but never used. I just made a new commit. Can you test it again?

chtinnes commented:
@eydam-prototyping: Yes, with the new commit, it's working now for me. Thanks.

chtinnes commented:
I just tried to use the LAN in an uasyncio task. It looks like the sockets are only working in an uasyncio task if the lan connection has been established in the same task. 
Is there an easy fix to this?

eydam-prototyping commented:
I don't have much experience with uasyncio. Could you provide a minimal example? Maybe I find some time to test it.

chtinnes commented:
@eydam-prototyping : the previous comment was not right. It turned out to be another problem not related to LAN.

eydam-prototyping commented:
@dpgeorge : Do I have to do anything else to get this PR merged?

xorbit commented:
I have confirmed that this works with both LAN8720 (current production) and RTL8201 (next rev prototype) on a [wESP32](https://wesp32.com/).

Since we are now without up-to-date MicroPython with support for Ethernet, I would _really_ love to see this merged before the next release.  Any reason it isn't?  It seems to be working great.

mirozbiro commented:
+1 for the merge as xorbit is asking.

spacemanspiff2007 commented:
+1 for the merge because I have to use the latest version because I am using a neopixel on pin>30 and now I can't use the ethernet connection.

dpgeorge commented:
Sorry for the delay.  I did not have any appropriate hardware to test this with so thanks to all who did test it.

Squashed and merged in 48437cec4547a21af73102c1ccdf079f63fff534 with some minor changes to the exception messages to make them more consistent with existing code, and also a change to the guard macros so it will compile with IDF 4.1.

Thanks @eydam-prototyping !

luke-jr commented:
Why did this remove the ability to set `clock_mode`? :/

dpgeorge commented:
> Why did this remove the ability to set `clock_mode`? :/

I think because it's no longer supported by the IDF, to dynamically select `clock_mode`.

eydam-prototyping commented:
Yes, that's right. Currently you have to set the clock mode in ESP-IDF at compile time. But as far as I know the guys at ESP-IDF are working on a solution.

dpgeorge commented:
It looks like the ability to dynamically select the clock was added in the IDF in https://github.com/espressif/esp-idf/commit/a6db3a5e77d92323e86c9c4d4dcf9e9c319ed498

eydam-prototyping commented:
Ok, great. But we have to wait for the next release of ESP-IDF, right? 
I'll make a new PR when it is released.

rcastberg commented:
Do we have any idea when a supported ESP-IDF will be released with support for clock_mode?

steeley commented:
Is this actually included in v1.18?
esp32 dev kit board wired like this:
https://sautter.com/blog/ethernet-on-esp32-using-lan8720/

running the code above just gives:
`Initializing LAN
Traceback (most recent call last):
  File "<stdin>", line 6, in <module>
OSError: esp_eth_set_default_handlers failed (invalid parameter)`
